File: src/services/vton.service.ts
import { Client } from "@gradio/client";
import type {
	VTONRequest,
	VTONResponse,
	GradioStatusEvent,
	GradioDataEvent,
	ProgressInfo,
} from "../types/index.js";

export type ProgressCallback = (info: ProgressInfo) => void;

export class VTONService {
	private readonly spaceId = "m1ns2o/IDM-VTON";
	private readonly apiName = "/tryon";
	// Estimated processing time in seconds (based on typical IDM-VTON runs)
	private readonly ESTIMATED_PROCESSING_TIME = 30;

	async processTryOn(
		request: VTONRequest,
		onProgress?: ProgressCallback
	): Promise<VTONResponse> {
		// Track state for accurate progress
		let lastKnownEta: number | undefined;
		let lastQueuePosition: number | undefined;
		let lastQueueSize: number | undefined;
		let processingStartTime: number | undefined;
		let progressUpdateInterval: ReturnType<typeof setInterval> | undefined;

		const sendProgress = (info: Partial<ProgressInfo>) => {
			onProgress?.({
				status: info.status ?? "processing",
				progress: info.progress ?? 0,
				message: info.message ?? "Processing...",
				eta: info.eta ?? lastKnownEta,
				queuePosition: info.queuePosition ?? lastQueuePosition,
				queueSize: info.queueSize ?? lastQueueSize,
				stepProgress: info.stepProgress,
			});
		};

		try {
			console.log("Connecting to Gradio space:", this.spaceId);
			sendProgress({
				status: "connecting",
				progress: 5,
				message: "Connecting to Gradio space...",
			});

			const hfToken = process.env.HF_TOKEN;
			if (hfToken) {
				console.log("Using Hugging Face authentication token");
			} else {
				console.log("No HF token found, connecting without authentication");
			}

			const app = await Client.connect(this.spaceId, {
				hf_token: hfToken as `hf_${string}` | undefined,
			});

			const humanImageBlob = new Blob([request.humanImage], {
				type: "image/png",
			});
			const garmentImageBlob = new Blob([request.garmentImage], {
				type: "image/png",
			});

			const imageEditorData = {
				background: humanImageBlob,
				layers: [],
				composite: null,
			};

			console.log("Sending request to Gradio API...");
			sendProgress({
				status: "submitting",
				progress: 10,
				message: "Submitting request to server...",
			});

			const job = app.submit(this.apiName, [
				imageEditorData,
				garmentImageBlob,
				request.description || "A person wearing the garment",
				request.autoMask ?? true,
				request.autoCrop ?? true,
				request.denoisingSteps ?? 30,
				request.seed ?? 42,
			]);

			// Start progress interval immediately after job submission
			processingStartTime = Date.now();
			progressUpdateInterval = setInterval(() => {
				if (!processingStartTime) return;

				const elapsed = (Date.now() - processingStartTime) / 1000;
				const estimatedTotal = this.ESTIMATED_PROCESSING_TIME;

				// Progress from 15% to 95% based on elapsed time
				// Use easing function for more natural feel
				const rawProgress = elapsed / estimatedTotal;
				const easedProgress = 1 - Math.pow(1 - Math.min(rawProgress, 1), 2);
				const progressPercent = 15 + easedProgress * 80;

				const remainingTime = Math.max(0, Math.ceil(estimatedTotal - elapsed));

				sendProgress({
					status: "generating",
					progress: Math.min(95, progressPercent),
					message: `Generating... (~${remainingTime}s remaining)`,
					eta: remainingTime,
				});
			}, 500);

			// Use Promise wrapper with async IIFE to handle iterator properly
			return new Promise<VTONResponse>((resolve, reject) => {
				(async () => {
					try {
						for await (const event of job) {
							console.log("Job event:", JSON.stringify(event));

							if (event.type === "status") {
								const statusEvent = event as unknown as GradioStatusEvent;
								const stage = statusEvent.stage;

								if (statusEvent.eta !== undefined) {
									lastKnownEta = statusEvent.eta;
								}
								if (statusEvent.position !== undefined) {
									lastQueuePosition = statusEvent.position;
								}
								if (statusEvent.queue_size !== undefined) {
									lastQueueSize = statusEvent.queue_size;
								}

								if (stage === "pending") {
									const positionText =
										lastQueuePosition !== undefined
											? `Position ${lastQueuePosition + 1}${lastQueueSize ? ` of ${lastQueueSize}` : ""}`
											: "Waiting in queue";
									const etaText = lastKnownEta
										? ` (ETA: ${Math.ceil(lastKnownEta)}s)`
										: "";

									sendProgress({
										status: "pending",
										progress: 15,
										message: `${positionText}${etaText}...`,
										queuePosition: lastQueuePosition,
										queueSize: lastQueueSize,
										eta: lastKnownEta,
									});
								} else if (stage === "generating") {
									// Progress is handled by the main interval
								} else if (stage === "complete") {
									// Just wait for data event
								} else if (stage === "error") {
									// Clear progress interval
									if (progressUpdateInterval) {
										clearInterval(progressUpdateInterval);
										progressUpdateInterval = undefined;
									}
									sendProgress({
										status: "error",
										progress: 0,
										message: statusEvent.message || "An error occurred",
									});
								}
							} else if (event.type === "data") {
								// Clear progress interval
								if (progressUpdateInterval) {
									clearInterval(progressUpdateInterval);
									progressUpdateInterval = undefined;
								}

								const dataEvent = event as unknown as GradioDataEvent;

								sendProgress({
									status: "complete",
									progress: 100,
									message: "Done!",
								});

								console.log("Received response from Gradio API");
								console.log(
									"Response data:",
									JSON.stringify(dataEvent.data, null, 2)
								);

								if (!dataEvent.data || dataEvent.data.length < 2) {
									reject(new Error("Invalid response from Gradio API"));
									return;
								}

								const outputImage =
									typeof dataEvent.data[0] === "object" &&
									dataEvent.data[0] !== null
										? (dataEvent.data[0] as any).url
										: dataEvent.data[0];

								const maskedImage =
									typeof dataEvent.data[1] === "object" &&
									dataEvent.data[1] !== null
										? (dataEvent.data[1] as any).url
										: dataEvent.data[1];

								console.log("Output image URL:", outputImage);
								console.log("Masked image URL:", maskedImage);
								console.log("Returning successful response");

								resolve({
									success: true,
									outputImage,
									maskedImage,
								});
								return; // Exit the async function immediately after resolve
							}
						}
						// If loop finishes without data
						if (progressUpdateInterval) {
							clearInterval(progressUpdateInterval);
							progressUpdateInterval = undefined;
						}
						reject(new Error("Stream finished without returning data"));
					} catch (err) {
						if (progressUpdateInterval) {
							clearInterval(progressUpdateInterval);
							progressUpdateInterval = undefined;
						}
						reject(err);
					}
				})();
			});
		} catch (error) {
			console.error("Error processing virtual try-on:", error);
			sendProgress({
				status: "error",
				progress: 0,
				message: error instanceof Error ? error.message : "Unknown error",
			});
			return {
				success: false,
				error:
					error instanceof Error ? error.message : "Unknown error occurred",
			};
		}
	}
}

File: backend/app/services/vton_service.py
"""
Virtual Try-On Service
Gradio API를 사용하여 가상 의류 착용을 수행합니다.
"""

import asyncio
import os
import time
import logging
from typing import Callable, Optional
from dataclasses import dataclass
from gradio_client import Client, handle_file
import tempfile
import base64

# 로깅 설정
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


@dataclass
class ProgressInfo:
    """진행 상태 정보"""
    status: str  # connecting, submitting, pending, generating, complete, error
    progress: float
    message: str
    eta: Optional[float] = None
    queue_position: Optional[int] = None
    queue_size: Optional[int] = None


@dataclass
class VTONRequest:
    """Virtual Try-On 요청"""
    human_image: bytes
    garment_image: bytes
    description: str = "A person wearing the garment"
    auto_mask: bool = True
    auto_crop: bool = True
    denoising_steps: int = 30
    seed: int = 42


@dataclass
class VTONResponse:
    """Virtual Try-On 응답"""
    success: bool
    output_image: Optional[str] = None
    masked_image: Optional[str] = None
    error: Optional[str] = None


ProgressCallback = Callable[[ProgressInfo], None]


class VTONService:
    """Virtual Try-On 서비스"""

    def __init__(self):
        self.space_id = "m1ns2o/AI-Clothes-Changer"
        self.api_name = "/infer"
        self.estimated_processing_time = 30  # seconds

    async def process_tryon(
        self,
        request: VTONRequest,
        on_progress: Optional[ProgressCallback] = None
    ) -> VTONResponse:
        """Virtual Try-On 처리"""

        def send_progress(
            status: str,
            progress: float,
            message: str,
            eta: Optional[float] = None,
            queue_position: Optional[int] = None,
            queue_size: Optional[int] = None
        ):
            if on_progress:
                on_progress(ProgressInfo(
                    status=status,
                    progress=progress,
                    message=message,
                    eta=eta,
                    queue_position=queue_position,
                    queue_size=queue_size
                ))

        try:
            logger.info(f"Connecting to Gradio space: {self.space_id}")
            send_progress("connecting", 5, "Connecting to Gradio space...")

            # HF Token 확인
            hf_token = os.environ.get("HF_TOKEN")
            if hf_token:
                logger.info("Using Hugging Face authentication token")
            else:
                logger.info("No HF token found, connecting without authentication")

            # Gradio Client 연결 (verbose=False로 로그 줄임)
            client = Client(self.space_id, token=hf_token, verbose=False)
            # heartbeat 비활성화 (404 에러 방지)
            if hasattr(client, '_kill_heartbeat'):
                client._kill_heartbeat.set()

            # 임시 파일로 이미지 저장
            with tempfile.NamedTemporaryFile(suffix=".png", delete=False) as human_file:
                human_file.write(request.human_image)
                human_path = human_file.name

            with tempfile.NamedTemporaryFile(suffix=".png", delete=False) as garment_file:
                garment_file.write(request.garment_image)
                garment_path = garment_file.name

            logger.info("Sending request to Gradio API...")
            send_progress("submitting", 10, "Submitting request to server...")

            # 진행 상황 업데이트를 위한 백그라운드 태스크
            processing_start_time = time.time()
            progress_task_running = True

            async def update_progress():
                while progress_task_running:
                    elapsed = time.time() - processing_start_time
                    estimated_total = self.estimated_processing_time

                    # 15% ~ 95% 진행률 계산 (easing 적용)
                    raw_progress = elapsed / estimated_total
                    eased_progress = 1 - (1 - min(raw_progress, 1)) ** 2
                    progress_percent = 15 + eased_progress * 80

                    remaining_time = max(0, int(estimated_total - elapsed))

                    send_progress(
                        "generating",
                        min(95, progress_percent),
                        f"Generating... (~{remaining_time}s remaining)",
                        eta=remaining_time
                    )
                    await asyncio.sleep(0.5)

            # 백그라운드에서 진행률 업데이트 시작
            progress_task = asyncio.create_task(update_progress())

            try:
                # Gradio API 호출 (블로킹 호출을 스레드풀에서 실행)
                loop = asyncio.get_event_loop()
                result = await loop.run_in_executor(
                    None,
                    lambda: client.predict(
                        person=handle_file(human_path),
                        garment=handle_file(garment_path),
                        denoise_steps=request.denoising_steps,
                        seed=request.seed,
                        api_name=self.api_name
                    )
                )
            finally:
                # 진행률 업데이트 중지
                progress_task_running = False
                progress_task.cancel()
                try:
                    await progress_task
                except asyncio.CancelledError:
                    pass

                # 임시 파일 정리
                try:
                    os.unlink(human_path)
                    os.unlink(garment_path)
                except:
                    pass

            send_progress("complete", 100, "Done!")

            logger.info("Received response from Gradio API")
            logger.info(f"Response data: {result}")

            # 결과 처리 (frogleo/AI-Clothes-Changer는 단일 이미지 반환)
            if result:
                output_path = result if isinstance(result, str) else None
                logger.info(f"Output image path: {output_path}")

                # 파일을 읽어서 base64로 변환
                output_image_base64 = None
                if output_path and os.path.exists(output_path):
                    with open(output_path, "rb") as f:
                        image_data = f.read()
                        output_image_base64 = f"data:image/png;base64,{base64.b64encode(image_data).decode('utf-8')}"
                    # 임시 파일 정리
                    try:
                        os.unlink(output_path)
                    except:
                        pass

                return VTONResponse(
                    success=True,
                    output_image=output_image_base64,
                    masked_image=None
                )
            else:
                return VTONResponse(
                    success=False,
                    error="Invalid response from Gradio API"
                )

        except Exception as e:
            logger.error(f"Error processing virtual try-on: {e}")
            send_progress("error", 0, str(e))
            return VTONResponse(
                success=False,
                error=str(e)
            )
